### PR TITLE
Use java 9 compiler, continue targeting java 8

### LINKED ISSUE #1882: Make TripleA java-9 ready
I managed to get gradle compiling TripleA using the lates java 9 jdk-preview:

Current gradle output when compiling against java 9:
[Gist](https://gist.github.com/RoiEXLab/bcc46a53c62958a4bb472a0c4b3f5e8f#file-java9)
Output when compiling with the latest java 9 jdk with java 8 source setting:
[Gist](https://gist.github.com/RoiEXLab/bcc46a53c62958a4bb472a0c4b3f5e8f#file-java9-8)

Some errors are java 9 specific (due to the new module system), we don't need to worry about them right now, but most errors can be "fixed" while still being able to compile TripleA with java 8.
I will create a PR adressing a couple of those issues...
(I will also update the log to representr the state of current master)

[Java 9 Countdown](http://www.java9countdown.xyz/)

EDIT: Updated log to reflect the state after #1883 gets merged

### PR DESCRIPTION
**Pull request by RoiEXLab**
Closes #1882
Interestingly enough upgrading resulted in test failures (all of the same kind) that really should have happened with JDK 8 already.
Basically `MutedMacController#getMacUnmuteTime` and `MutedUsernameController#getUsernameUnmuteTime` return a milisecond-precision value that is compared against a nanosecond-precision Instant.
I workaround this issue by simply ignoring the nanoseconds, but I will submit a follow-up PR adressing this issue the right way, by fixing this instead of working around it.
I just didn't want to increase the scope of this small PR for a change that requires quite some refactoring.

This change is fully backwards compatible, all of the binaries should (if the gradle passes source and target correctly to the compiler) still be executable on JREs >= 8
Install4j supports Java 9, postgres triggers a reflection warning, that should be fixed in the next driver release, gradle itself triggers such a warning as well, but we can be pretty confident this will be fixed as well ^^

### CODE REVIEW

**Comment by ssoloff** [comment]
@RoiEXLab I don't have a problem with the upgrade, but I think we all need to synchronize on this so we don't have issues that pop up on Travis but not in a local build for a dev who may still be using JDK 8.  That is, the entire dev team should be ready to switch to JDK 9 once this PR is merged.

Also, I could have sworn there was more detailed dev environment setup documentation when I first joined this project, but I can't seem to find it anymore under _docs/dev/_.  Would this be a good PR to add some minimal instructions (e.g. which JDK to install)?

File: src/integ_test/java/games/strategy/engine/lobby/server/db/MutedMacControllerIntegrationTest.java [hunk 1]
```diff
@@ -59,8 +59,12 @@ public void testMuteMacUpdate() {
   }
 
   private Instant muteMacForSeconds(final long length) {
-    final Instant muteEnd = length == Long.MAX_VALUE ? null : Instant.now().plusSeconds(length);
+    final Instant muteEnd = length == Long.MAX_VALUE ? null : nanosIgnored(Instant.now().plusSeconds(length));
     controller.addMutedMac(hashedMac, muteEnd);
     return muteEnd;
   }
+  
+  private Instant nanosIgnored(final Instant precise) {
```
Comment (ssoloff) [inline src/integ_test/java/games/strategy/engine/lobby/server/db/MutedMacControllerIntegrationTest.java]: I know you said you plan on removing these workarounds in a follow-up PR, but just in case that gets delayed for some reason, a big fat FIXME comment referencing this PR might be a good idea.

If, for some reason, you think these methods may exist for some time, a better name might be `withoutNanos()`.

Nit: These methods can be static.

File: src/integ_test/java/games/strategy/engine/lobby/server/db/MutedMacControllerIntegrationTest.java [hunk 2]
```diff
@@ -59,8 +59,12 @@ public void testMuteMacUpdate() {
   }
 
   private Instant muteMacForSeconds(final long length) {
-    final Instant muteEnd = length == Long.MAX_VALUE ? null : Instant.now().plusSeconds(length);
+    final Instant muteEnd = length == Long.MAX_VALUE ? null : nanosIgnored(Instant.now().plusSeconds(length));
```
Comment (ssoloff) [inline src/integ_test/java/games/strategy/engine/lobby/server/db/MutedMacControllerIntegrationTest.java]: Just a suggestion for the "real" fix...  Having tests depend on wall time (i.e. `Instant#now()`) is generally a Bad Thing.  We should try to refactor these tests to use fixed `Instant`s of our choosing.  That probably means we need to be able to inject an `Instant` into the controller because I'm guessing it's using `Instant#now()`, as well.

**Comment by prastle** [comment]
@ssoloff  off topic just notice u posted... please talk to @RoiEXLab  and @ron-murhammer  and @DanVanAtta  about the certifications it seems possible now.

**Comment by ssoloff** [comment]
@prastle I'm lacking some context here. :smile: What certifications are we talking about?

**Comment by prastle** [comment]
Comodo certificate

**Comment by ssoloff** [comment]
@prastle Ah, ok.  Last I heard, you and @RoiEXLab were working on that.  I don't remember being involved. :smile:

**Comment by prastle** [comment]
Simple answer is Yes! But anyone not in Canada in driving distance needs to be notarized and faxed.

**Comment by RoiEXLab** [comment]
Blocked until #2755 gets merged

**Comment by RoiEXLab** [comment]
Rebased

**Comment by ssoloff** [comment]
One additional thing we need to consider...  If devs upgrade to JDK 9, then, presumably, most (if not all) of our dev testing will be done on Java 9, and we may miss something that only crops up in Java 8 (the platform a lot of users will still be using for a while).  The obvious solution is to run a matrix build with Travis using both JDK 8 and JDK 9 so at least the unit and integration tests are run in both environments.  

However, I'm not sure how that will affect releases.  We obviously don't want two releases per push to `master`.  The only way I can think to get around that at the moment is to define an environment variable as part of the matrix that is checked during the release phase.  And then we would only set one of those environment variables to "true" which indicates a release should be performed.

Also, IIRC, once you enable a matrix build, the Travis build numbers start looking like "\<build\>.\<job\>", e.g. 7621.1, 7621.2, etc.  That could screw up our version numbers if we don't trim the job number off the end.

**Comment by RoiEXLab** [comment]
@ssoloff Good Idea, however I'm not 100% convinced it would we worth the additional work.
We have problems with outdated JREs already and I don't think it's going to make a big difference running tests on an older JRE.
Even if we're going to run into any issues with this couldn't we just tell users to update their java version?

**Comment by ssoloff** [comment]
> We have problems with outdated JREs already and I don't think it's going to make a big difference running tests on an older JRE.

I would argue there's a big difference between running tests on different minor revisions of the JRE vs. running tests on different major revisions of the JRE.  Especially given the significant changes between Java 8 and 9 (as opposed to, say, the relatively minor changes between Java 6 and 7).  But I agree with you that it may not be worth the amount of work involved (it would be a no-brainer if it wasn't for the Travis release phase).

> Even if we're going to run into any issues with this couldn't we just tell users to update their java version?

Yup, that's always a possibility.  My only concern is the type of first-time user who uses Java 8, experiences a problem in TripleA that we didn't know about because we're not testing with Java 8, doesn't bother to report it, and just gives up and uninstalls TripleA.

**Comment by RoiEXLab** [comment]
@ssoloff
> I would argue there's a big difference between running tests on different minor revisions of the JRE vs. running tests on different major revisions of the JRE. Especially given the significant changes between Java 8 and 9[...].

I'd agree with you in general, but in this case we're still compiling Java 8 compatible byte code, so any of the "big changes" can't be really used.
The only thing that changes is that we (at least AFAIK) get to use diamond notation on anonymous classes and some optimizations that were made to the compiler in regards to string concatenation.

>  My only concern is the type of first-time user who uses Java 8, experiences a problem in TripleA that we didn't know about because we're not testing with Java 8, doesn't bother to report it, and just gives up and uninstalls TripleA.

Good Point. I think this is a rather general issue that deserves a dedicated issue on its own.
Long story short, we probably want to implement something that gives the user exact steps on how to proceed when TripleA crashes.
Not something as big as for example the Minecraft Hopper system, where the exception is uploaded, analysed, compared against a database and gives the user information on how to fix the issue by updating their drivers or something, but rather something simple as: Try to update java and if it still doesn't work go to github or the forum and ask for help there.

**Comment by RoiEXLab** [comment]
Any opinions on this @DanVanAtta @ron-murhammer ?

**Comment by ron-murhammer** [comment]
I'm alright with this. The key is that we still must target 8 compatibility for a while since most users won't upgrade to 9 for some time.

**Comment by DanVanAtta** [comment]
@RoiEXLab what are your thoughts on building for both 8 and 9 platforms?

eg:
```
 jdk:
  - oraclejdk8
  - oraclejdk9
```

I think it could have the benefit that we would be sure the jdk8 build was good, and we would also get notification if we did something to break java9

**Comment by ssoloff** [comment]
@DanVanAtta 

> what are your thoughts on building for both 8 and 9 platforms?

I've been trying to test this on my fork, but the Travis build failures aren't making it easy.  I've only gotten one good build all day. 😞

**Comment by RoiEXLab** [comment]
@DanVanAtta no objections on that suggestion

**Comment by ssoloff** [comment]
Now that Travis is operational again, I was able to complete the tests I had intended to run today.

Just doing what @DanVanAtta proposed (i.e. having two entries in the `jdk` list) appears to work, but there is a race condition.  Both build jobs create a release on GitHub, but only the last one to run wins.  This works without raising an error because our build scripts delete any existing tag before deploying.

As suggested above, I think this can be easily resolved by simply adding an environment variable as part of the build matrix that enables or disables the release phase for a particular job.  I'll work on that tomorrow.  I'm assuming we want to release the Java 9 build.  Please advise if the Java 8 build should be released instead.

**Comment by ssoloff** [comment]
I submitted RoiEXLab/triplea#9 which enables a matrix build on Travis for both Java 8 and Java 9.  The releases will come from the Java 9 build.  Please advise if the Java 8 build should instead be used to generate releases.

**Comment by RoiEXLab** [comment]
@ssoloff Did you check if the TRAVIS build number doesn't break the versioning system?

**Comment by ssoloff** [comment]
@RoiEXLab Yes.  `TRAVIS_BUILD` only contains the build number.  The job number is stored separately in `TRAVIS_JOB`.  That's why if both jobs attempt to release, the latter overwrites the results of the former.

I built this change and generated a release on my fork.  The Travis build is [here](https://travis-ci.org/ssoloff/triplea-game-triplea/builds/324016812), and the release is [here](https://github.com/ssoloff/triplea-game-triplea/releases/tag/ssoloff-1.9.0.0.1625).  (Note that the long build time was simply due to latent issues from yesterday's Travis build problems.)

I also smoke tested my custom release and made sure I could start a new local game.

**Comment by RoiEXLab** [comment]
Ok leaving up to one of you to merge this then.

**Comment by ssoloff** [comment]
Since @ron-murhammer already approved, I'll go ahead and merge.  I believe @DanVanAtta's (and my) concern has been addressed.  @DanVanAtta, if you'd prefer that the Java 8 build generate the releases, please say so, and we can change that quickly with a two-line change.

**Comment by ssoloff** [comment]
Whoops.  I mistakenly copy-pasted a `+` on line 4, which is preventing Travis from recognizing the matrix build configuration.  I'll fix ASAP.

**Comment by ssoloff** [comment]
Fixed in #2795.

**Comment by ssoloff** [comment]
Build 8217 is the first release built with Java 9.  I installed the Windows x64 version on a machine with only Java 8 installed and confirmed I could start a new local game.
